reposerver/repository/repository.go

```diff
@@ -1253,6 +1362,259 @@ func getRepoCredential(repoCredentials []*v1alpha1.RepoCreds, repoURL string) *v
+				logCtx.Errorf("copyFilesystemSubtree: filepath.Base failed: %v", err)
+				return fmt.Errorf("Unable to extract source file's name")
+			}
+
+			//dstFile = filepath.Join(dstDir, fname)
+			dstFile, err = secpath.SecureJoin(dstDir, fname)
+			if err != nil {
+				logCtx.Errorf("SecJoin(%s, %s) failed: %v", dstDir, fname, err)
+				return err
+			}
+		}
+		// Now copy the file over
+		if err := argoutil.CopyFile(srcFullPath, dstFile, 0644); err != nil {
+			logCtx.Errorf("copyFilesystemSubtree: CopyFile failed: %v", err)
+			return err
+		}
+	} else if srcStat.Mode().IsDir() {
+		// now we are copying a directory
+
+		// Make sure that the destination directory exists
+		if err := os.MkdirAll(dstFullPath, 0755); err != nil {
+			logCtx.Debugf("copyFilesystemSubtree: os.MkdirAll failed: %v", err)
+			return err
+		}
+
+		// Copy the content of the source directory to the destination dir
+		err = filepath.WalkDir(srcFullPath, func(path string, d fs.DirEntry, err error) error {
+			return copyFilesystemSubtreeWalker(srcFullPath, dstFullPath, failOnOutOfBoundsSymlink, path, d, err)
+		})
+		if err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("Copy src is neither a file or a directory")
+	}
+	return nil
+}
+
+func copyFilesystemSubtreeWalker(srcFullPath string, dstFullPath string, failOnOutOfBoundsSymlink bool, path string, d fs.DirEntry, err error) error {
+	// we propagate encountered errors
+	if err != nil {
+		return err
+	}
+	logCtx := log.WithFields(log.Fields{
+		"srcFullPath": srcFullPath,
+		"dstFullPath": dstFullPath,
+		"path": path,
+	})
+
+	logCtx.Debug("CopyFrom processing entry")
+
+	// first get the relative path of the current entry to srcFullPath:
+	relpath, err := filepath.Rel(srcFullPath, path)
+	if err != nil {
+		logCtx.Errorf("CopyFrom cannot get relpath: %v", err)
+		return err
+	}
+
+	//targetPath := filepath.Join(dstFullPath, relpath)
+	targetPath, err := secpath.SecureJoin(dstFullPath, relpath)
+	if err != nil {
+		logCtx.Errorf("SecJoin(%s, %s) failed: %v", dstFullPath, relpath, err)
+		return err
+	}
+	logCtx = logCtx.WithFields(log.Fields{
+		"targetPath": targetPath,
+		"relpath": relpath,
+	})
+
+	// if it's a directory, then we have to create it
+	// files are to be copied
+	// symlinks checked
+	// rest ignored
+	if d.IsDir() {
+		if err := os.Mkdir(targetPath, 0755); err != nil && dstFullPath != targetPath {
+			logCtx.Errorf("CopyFrom mkdir failed: %v", err)
+			return err
+		}
+		logCtx.Debug("CopyFrom directory created")
+		return nil
+	} else if d.Type().IsRegular() {
+		// regular files are copied with CopyFile.
+		// Since it's a linear recursive walk on the fs, we can expect
+		// the parent directories to exist already
+
+		if err = argoutil.CopyFile(path, targetPath, 0644); err != nil {
+			logCtx.Errorf("CopyFrom CopyFile failed: %v", err)
+			return err
+		}
+		logCtx.Debug("CopyFrom file copied")
+		return nil
+	} else if d.Type()&os.ModeSymlink>0 {
```

Comment:
nit: `d.Type()&os.ModeSymlink>0` -> `d.Type() & os.ModeSymlink > 0`

Reply:
`19d8f891` fixes this.